### DISCUSSION TITLE
Question about Cloudflare Workers support

### DISCUSSION THREAD

**Discussion by elct9620** [fedify-dev/fedify#204] [Q&A] [answered]
Hi,

I find the [microblog document](https://fedify.dev/tutorial/microblog) has some notes about `Recently, cloud edge functions like Cloudflare Workers have also gained popularity as JavaScript runtimes` 

Does Fedify expect to support it natively?

If I choose to use Cloudflare Workers, I need to use the [Bindings](https://developers.cloudflare.com/pages/functions/bindings/#kv-namespaces) feature to get the key-value store.

```ts
interface Bindings {
  Cache: KVNamespace;
}

const app = new Hono<{ Bindings: Bindings }>()

// ...
app.use(federation(fedi, (ctx) => {
   // the "ctx.env.Cache" is here
  return // ...
}));  
```

However, the `Federation` object is required `kv` and initialized before handling the request.

```ts
const fedi = createFederation<string>({
  // ctx.env.Cache is not available yet
});

const app = new Hono();
app.use(federation(fedi, (ctx) => "context data"));  
```

Or the expected usage is like this

```ts
interface Bindings {
  Cache: KVNamespace;
}

const app = new Hono<{ Bindings: Bindings }>()

app.use(async (ctx, next) => {
  const fedi = createFederation<string>({
    kv: new MyKVStore(ctx.env.Cache)
  });
  
  ctx.set('fedi', fedi) // use this later

  await next()
});  


```

**Comment by dahlia** [comment]
Creating a `Federation` instance for every request technically makes no problem, but I doubt that Fedify currently work well on Cloudflare Workers, as it's never tested on that. 🤔

**Comment by elct9620** [comment]
I will test it before I start my project. If I encounter any problems, I will report them or make a PR.

**Comment by elct9620** [comment]
Tested with Cloudflare Workers (local dev mode) but the `shim-deno` broke it.

```
✘ [ERROR] service core:user:poc-fedify: Uncaught ReferenceError: __dirname is not defined

    at null.<anonymous> (index.js:4178:186) in
  node_modules/@deno/shim-deno/dist/deno/stable/variables/mainModule.js
    at null.<anonymous> (index.js:32:50) in __require2
    at null.<anonymous> (index.js:6211:27) in
  node_modules/@deno/shim-deno/dist/deno/stable/variables.js
    at null.<anonymous> (index.js:32:50) in __require2
    at null.<anonymous> (index.js:11061:26) in
  node_modules/@deno/shim-deno/dist/deno/stable/functions.js
    at null.<anonymous> (index.js:32:50) in __require2
    at null.<anonymous> (index.js:11503:18) in node_modules/@deno/shim-deno/dist/deno/stable/main.js
    at null.<anonymous> (index.js:32:50) in __require2
    at null.<anonymous> (index.js:11541:18) in node_modules/@deno/shim-deno/dist/deno.js
    at null.<anonymous> (index.js:32:50) in __require2
```

It seems some packages or features are not supported by Cloudflare Worker.
https://developers.cloudflare.com/workers/runtime-apis/nodejs/

Is it possible to avoid using shim in Fedify for better capabilities?

**Comment by dahlia** [comment]
We're willing to support edge functions including Cloudflare Workers in a long term, but I guess it wouldn't be trivial in a short term… 😭 One cloud edge function I've tested with Fedify is Deno Deploy at the moment.

**Comment by dahlia** [comment]
Ah, [Fedify on JSR](https://jsr.io/@fedify/fedify) would work better with edge functions.

**Comment by elct9620** [comment]
Switching to `npx jsr add @fedify/fedify` can resolve the shim issue. However, it throws a `Temporal is not defined` error.

> I use `new SimpleKvStore()` here, but not relate it because I created a mock store that still does not work.

```
✘ [ERROR] ReferenceError: Temporal is not defined

      at kvCache
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/@fedify/fedify/runtime/docloader.ts:356:26)
      at new FederationImpl
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/@fedify/fedify/federation/middleware.ts:354:53)
      at createFederation
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/@fedify/fedify/federation/middleware.ts:245:10)
      at Array.<anonymous> (file:///Users/elct9620/Desktop/poc-fedify/src/index.ts:7:16)
      at Object.dispatch
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/hono/dist/hono-base.js:187:37)
      at Object.fetch
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/hono/dist/hono-base.js:213:17)
      at fetchDispatcher
  (file:///Users/elct9620/Desktop/poc-fedify/.wrangler/tmp/bundle-ZHjIs9/middleware-loader.entry.ts:54:17)
      at __facade_invokeChain__
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/wrangler/templates/middleware/common.ts:53:9)
      at Object.next
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/wrangler/templates/middleware/common.ts:50:11)
      at jsonError
  (file:///Users/elct9620/Desktop/poc-fedify/node_modules/wrangler/templates/middleware/middleware-miniflare3-json-error.ts:22:30)


```

**Comment by elct9620** [comment]
Updated:

Fixed, the typo caused the web finger not to return anything.

---

I can run it correctly after adding polyfill.

According to the tutorial - https://fedify.dev/tutorial/basics#actor-dispatcher

I can call `curl -H "Accept: application/activity+json" http://localhost:8787/users/me` to get the correct response, but Web Finger (`curl "http://localhost:8787/.well-known/webfinger?resource=acct:me@localhost:887"`) returns 404.

**Comment by dahlia** [comment]
Could you try a [polyfill for Temporal API](https://github.com/fullcalendar/temporal-polyfill)?

**Comment by scruffymongrel** [comment]
> I can call `curl -H "Accept: application/activity+json" http://localhost:8787/users/me` to get the correct response, but Web Finger (`curl "http://localhost:8787/.well-known/webfinger?resource=acct:me@localhost:887"`) returns 404.

@elct9620, shouldn't `resource=acct:me@localhost:887` be `resource=acct:me@localhost:8787`?

Also, did you manage to get Fedify working with Cloudflare Workers?

**Comment by valtlai** [comment]
I got Fedify working on Cloudflare Workers by
- using the JSR version (as suggested earlier in this thread)
- adding the Temporal polyfill (as suggested earlier in this thread)
- stripping import attributes from the `.js` files in the `@fedify/fedify` module.

For the last one, I used this dirty solution:

```diff
// package.json

{
  "scripts": {
+    "prepare": "node ./patch-fedify-for-workers.js"
  }
}
```

```js
// patch-fedify-for-workers.js

import { glob, readFile, writeFile } from 'node:fs/promises';
import { join } from 'node:path';

const cwd = 'node_modules/@fedify/fedify';
const codeToDelete = ' with {\n  type: "json"\n}';

for await (const entry of glob('*/*.js', { cwd })) {
  (async () => {
    const filePath = join(cwd, entry);
    const originalContent = await readFile(filePath, 'utf8');
    const patchedContent = originalContent.replace(codeToDelete, '');
    if (originalContent !== patchedContent) {
      await writeFile(filePath, patchedContent);
    }
  })();
}
```

The `prepare` lifecycle script automatically runs after `npm install`.

This could be done better with esbuild, but I think this is good enough for myself as a temporary solution. I assume Workers will eventually support import attributes (and the Temporal API as well).

**Comment by valtlai** [comment]
The upcoming major version&nbsp;4 of wrangler supports import attributes. You can install a prerelease version by running `npm install wrangler@next`.[^1] Then you can remove the above npm lifecycle script.

[^1]: [cloudflare/workers-sdk#7893 (comment)](https://github.com/cloudflare/workers-sdk/issues/7893#issuecomment-2688262162)

**Reply by valtlai** [reply]
It’s better to install `wrangler@v4-rc` instead of `wrangler@next`. The latter just broke my build.

**Comment by dahlia** [comment]
Good news! We've officially added Cloudflare Workers support to the Fedify roadmap. I've created a detailed issue to track our implementation plan:

- #233

The effort will be tackled in phases, including compatibility assessment, core adaptations for Workers' environment, KV store and message queue implementations, and finally integration with Cloudflare's ecosystem. This will be a substantial project that we'll break down into several sub-issues.

Thank you everyone who expressed interest in this feature. Your input from this discussion has been valuable in shaping our approach. We welcome community contributions and feedback as we move forward with implementation.

If you're interested in contributing to any specific aspect of Workers support, please comment on the main issue to coordinate efforts.

**Comment by dahlia** [comment]
First milestone reached! PR #239 successfully implements a comprehensive test environment for Cloudflare Workers using Miniflare. Our test suite (with some exclusions) now passes in the Workers runtime environment.

This establishes the foundation for testing cross-runtime compatibility and validates that the core Fedify functionality can work on Workers. Next steps include addressing the remaining test failures and implementing Workers-specific components.

**Comment by dahlia** [comment, marked as answer]
🎉 Cloudflare Workers support is now complete! After implementing the test infrastructure, core module, examples, and comprehensive documentation, Fedify can now run on Cloudflare Workers.

**What's included:**

- `@fedify/fedify/x/cfworkers` module with [`WorkersKvStore`](https://unstable.fedify.dev/manual/kv#workerskvstore-cloudflare-workers-only) and [`WorkersMessageQueue`](https://unstable.fedify.dev/manual/mq#workersmessagequeue-cloudflare-workers-only)
- [Complete documentation](https://unstable.fedify.dev/manual/deploy#cloudflare-workers) covering Workers-specific patterns and constraints
- [Working example](https://github.com/fedify-dev/fedify/tree/main/examples/cloudflare-workers?rgh-link-date=2025-06-01T06%3A20%3A37.000Z) deployed to production
- Integration with Cloudflare KV and Queues

**Try it now:** Available in the development release v1.6.1-dev.876+7b07d213:

- JSR: https://jsr.io/@fedify/fedify@1.6.1-dev.876+7b07d213
- npm: https://www.npmjs.com/package/@fedify/fedify/v/1.6.1-dev.876

This will be included in the upcoming Fedify 1.6 stable release. Thank you to everyone who requested this feature and provided feedback throughout the implementation!

**Comment by dahlia** [comment]
[Fedify 1.6.1, which supports Cloudflare Workers, was released.](https://github.com/fedify-dev/fedify/discussions/249)
